### PR TITLE
Refund update

### PR DESCRIPTION
**Pull request by BohdanLeasoft**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -276,11 +276,11 @@ public function refundOrder(string $order_id)
             throw new InvalidOrderStatusException($order->getStatus()->get(), 'completed');
         }
 
-        if (!$order->getCurrentTransaction()->isCaptured()) {
+        if ($order->getCurrentTransaction()->isCapturable() && !$order->getCurrentTransaction()->isCaptured()) {
             throw  new RefundFailedException('Order is not yet captured, only captured order could be refunded');
         }
 
-        return $this->api_client->refundOrder(id: $order_id, orderData: ['order_lines' => $order->getOrderLines()->toArray()]);
+        return $this->api_client->refundOrder(id: $order_id, orderData: ['amount' => $order->getAmount()->get(), "description" => "Order refund", 'order_lines' => $order->getOrderLines()->toArray()]);
```
Comment (BohdanLeasoft) [inline src/Client.php:283]: Refund for Klarna pay later return error: `The browser (or proxy) sent a request that this server could not understand.`
So, such extending of orderData help to solve this error

File: src/Entities/Order.php [hunk 2]
```diff
@@ -50,7 +50,7 @@ public function __construct(
         );
 
         if ($merchantOrderId) $this->merchantOrderId = $this->createSimpleField(
-            propertyName: 'merchantOrderId',
+            propertyName: 'merchant_order_id',
             value: $merchantOrderId
```
Comment (BohdanLeasoft) [inline src/Entities/Order.php:54]: Before this change, orders from drupal plugin hadn't merchant_order_id

File: src/Client.php [hunk 3]
```diff
@@ -264,23 +264,30 @@ public function getIdealIssuers(): IdealIssuers
...
 
-        return $this->api_client->refundOrder(id: $order_id, orderData: ['order_lines' => $order->getOrderLines()->toArray()]);
+        return $this->api_client->refundOrder(
+            id: $order_id,
+            orderData: [
+                'amount' => $amount->get() ?? $order->getAmount()->get(),
+                "description" => "Order refund",
+                'order_lines' => $order->getOrderLines()->toArray()
+            ]);
     }
```
Comment (BohdanLeasoft) [inline src/Client.php:291]: I also tested partial refund by drupal plugin. It's work

File: src/Client.php [hunk 4]
```diff
@@ -264,23 +264,30 @@ public function getIdealIssuers(): IdealIssuers
...
+    public function refundOrder(string $order_id, Properties\Amount $amount = null)
     {
         $order = $this->getOrder(id: $order_id);
 
         if ($order->getStatus()->get() != "completed") {
             throw new InvalidOrderStatusException($order->getStatus()->get(), 'completed');
         }
 
-        if (!$order->getCurrentTransaction()->isCaptured()) {
+        if ($order->getCurrentTransaction()->isCapturable() && !$order->getCurrentTransaction()->isCaptured()) {
```
Comment (tiutiunnykleasoft) [inline src/Client.php:280]: I think this condition better to divide by two cases, first one is when transaction is not captured, and next one is when transaction can't be captured. By this way we can return more valuable exception

File: src/Client.php [hunk 5]
```diff
@@ -264,23 +264,30 @@ public function getIdealIssuers(): IdealIssuers
      * Only completed orders could be refunded.
      *
      * @param string $order_id
+     * @param Properties\Amount $anount
```
Comment (tiutiunnykleasoft) [inline src/Client.php:267]: Add possible null value of the amount, and amount spelled incorrect there

File: src/Properties/VatPercentage.php [hunk 6]
```diff
@@ -37,10 +37,10 @@ public function __construct(Percentage|int $value)
...
     public function validate($value)
     {
-        if ($value < 1 || $value > 10000) throw new OutOfDiapasonException(
+        if ($value < 0 || $value > 10000) throw new OutOfDiapasonException(
             propertyName: $this->propertyName,
             value: $value,
-            min: 1,
+            min: 0,
             max: 10000);
     }
```
[the remark this answers is not in the document]
Comment (tiutiunnykleasoft) [reply src/Properties/VatPercentage.php:45]: 👍🏻

File: src/Client.php (hunk 3 above)
Comment (tiutiunnykleasoft) [reply src/Client.php:291]: 👍🏻

**Comment by tiutiunnykleasoft** [comment]
Add new version and updates to changelog

File: src/Client.php (hunk 4 above)
Comment (BohdanLeasoft) [reply src/Client.php:280]: "By this way we can return more valuable exception"
I don't think so. With two divided checks we will have exception when order doesn't require any capturing. 
To my mind this should work only for capturable orders

**Pull request closed without merging**